[bugfix] Accumulated_gradient and TensoBoard

tchaton opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
 Please also include relevant motivation and context.
 List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.
-->

This PR tries to improve the logging display on TensorBoard when using accumulated_grad_batches > 1. 
It also introduces `log_epoch_metrics_on_step` parameter within the trainer.

`log_epoch_metrics_on_step` idea was dropped !


Fixes #4304

``log_epoch_metrics_on_step = True``

<img width="449" alt="Screenshot 2020-11-18 at 11 38 55" src="https://user-images.githubusercontent.com/12861981/99529424-b3e71b00-2997-11eb-86da-c92b03697fc0.png">
<img width="1508" alt="Screenshot 2020-11-18 at 11 39 03" src="https://user-images.githubusercontent.com/12861981/99529432-b6497500-2997-11eb-8691-971d6ef4b126.png">


``log_epoch_metrics_on_step = False``

<img width="446" alt="Screenshot 2020-11-18 at 11 57 42" src="https://user-images.githubusercontent.com/12861981/99529441-b8abcf00-2997-11eb-9dc0-267091dc3647.png">
<img width="1055" alt="Screenshot 2020-11-18 at 11 57 58" src="https://user-images.githubusercontent.com/12861981/99529445-b9446580-2997-11eb-9f2f-e21245a80102.png">


## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @tchaton! Thanks for opening this PR. 

* In the file [`pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/47efe4266ab88fd09310cadc7c6566bd61fc752a/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py):

> [Line 108:121](https://github.com/PyTorchLightning/pytorch-lightning/blob/47efe4266ab88fd09310cadc7c6566bd61fc752a/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L108): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 120 characters)





Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

Borda commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -181,7 +181,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dic, step=None, is_train_step=False):
```
can we add docs what the`is_train_step` means?
```suggestion
    def log_metrics(self, metrics, grad_norm_dic, step=None, log_on_step=False):
```

Borda commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -271,7 +271,7 @@ def __init__(
 
             val_check_interval: How often to check the validation set. Use float to check within a training epoch,
-                use int to check every n steps (batches).
+                use int åto check every n steps (batches).
```
```suggestion
                use to check every n steps (batches).
```

edenlightning commented:
Ass discussed offline, i'd say this would be better as a param in log method than in the trainer.

edenlightning commented:
@tchaton can we split this to 2 PRs?

tchaton commented:
> @tchaton can we split this to 2 PRs?

Hey @edenlightning, I removed the parameters as you suggested. This PR contains only the fix for tensorboard logging in case of accumulated_gradient > 1

Borda commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -95,14 +95,14 @@ def cache_logged_metrics(self) -> Union[EpochResultStore, None]:
         self.configure_logger(logger)
         # todo: IDE is complaining, these shall be initialized in the Trainer init at leas as placeholders
-        #  and assign here the desired value
+        # and assign here the desired value
```
I think that the extra space is the to he parsed in IDE as single todo task

Borda commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -181,7 +181,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dic, step=None, log_train_step_metrics=False):
```
lets add it in docs and explain how it fuse with `step` argument
- what happen in `step=n` and `log_train_step_metrics=False`
- what happen in `step=None` and `log_train_step_metrics=True`

Borda replied:
shall it be `log_train_metrics_kind='step'|'epoch'`?

tchaton replied:
Hey @jirka, 
`log_train_step_metrics` is there only for internal reason. 
This should not be exposed in the doc or made over-complicated :)
I wanted to have a simple boolean to keep things fast.

Borda reviewed (commented):
sure about the APPI, and pls update docs

SkafteNicki commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -181,7 +181,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dic, step=None, log_train_step_metrics=False):
```
could you add description of the new arg under docstring?

tchaton replied:
done !

SkafteNicki reviewed (approved):
LGTM

tchaton commented:
Hey @williamFalcon , can you review this one ?
